### ISSUE TITLE
Out of bound read in mpls_print()

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#549]
**Out of bound read in mpls_print()** 

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r heap_bufover_3.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/19833309/ed50fec2-9e3d-11e6-8c72-83de78043354.png)

ASAN output:

```
=================================================================
==19172==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000ef36 at pc 0x0000005e4031 bp 0x7ffe02616c30 sp 0x7ffe02616c28
READ of size 1 at 0x60300000ef36 thread T0
    #0 0x5e4030 in mpls_print XYZ/tcpdump/./print-mpls.c:127:10
    #1 0x566f1c in ethertype_print XYZ/tcpdump/./print-ether.c:405:3
    #2 0x566418 in ether_print XYZ/tcpdump/./print-ether.c:229:7
    #3 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #4 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #5 0x7adaf7 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #6 0x6b6f4c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #7 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #8 0x7f064bc2182f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #9 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

0x60300000ef36 is located 0 bytes to the right of 22-byte region [0x60300000ef20,0x60300000ef36)
allocated by thread T0 here:
    #0 0x4b880c in malloc /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64:3
    #1 0x7af103 in pcap_check_header XYZ/libpcap/./sf-pcap.c:401:14
    #2 0x7acf22 in pcap_fopen_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:400:7
    #3 0x7accaf in pcap_open_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:307:6
    #4 0x7f064bc2182f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow XYZ/tcpdump/./print-mpls.c:127:10 in mpls_print
Shadow bytes around the buggy address:
  0x0c067fff9d90: 00 00 00 00 fa fa 00 00 00 01 fa fa 00 00 04 fa
  0x0c067fff9da0: fa fa 00 00 07 fa fa fa 00 00 00 01 fa fa 00 00
  0x0c067fff9db0: 00 02 fa fa 00 00 00 01 fa fa 00 00 06 fa fa fa
  0x0c067fff9dc0: 00 00 00 fa fa fa 00 00 07 fa fa fa 00 00 06 fa
  0x0c067fff9dd0: fa fa 00 00 07 fa fa fa 00 00 00 00 fa fa fd fd
=>0x0c067fff9de0: fd fa fa fa 00 00[06]fa fa fa 00 00 04 fa fa fa
  0x0c067fff9df0: 00 00 05 fa fa fa 00 00 04 fa fa fa 00 00 04 fa
  0x0c067fff9e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9e20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==19172==ABORTING
```

**Comment by fxlb** [comment]
Thanks for the report.
This issue is already fixed in the security branch (Will be merged in a few weeks).

**Comment by infrastation** [comment]
The original reporter for this vulnerability is Hanno Böck. It took a long time to publish the fix so the same issue was later independently discovered by other researchers.

Commit 69ead2a fixes this and the fix is available in the release 4.9.0. Thanks to everybody for their reports, please send any future security reports to security@tcpdump.org.
